### ISSUE TITLE
Attribute for skipping field of struct in Debug derives?

### ISSUE THREAD

**Issue by alexreg** [rust-lang/rust#37009]
Could we have something like this?

```rust
#[derive(Debug)]
struct Foo {
    pub a: A,
    pub b: B,
    #[skip_derive]
    pub c: C, // especially useful if C : !Debug
}
```

This could also feasibly be applied to other auto-derivations, of course (not just Debug).

**Comment by durka** [comment]
At minimum we could certainly do it with Macros 1.1 (then it would need a different name like `#[derive(DebugEx)]`).

**Comment by alexreg** [comment]
Yeah, that would be a step forward at least. Would be nice to have it work for the built-in Debug though, I think.

**Comment by mcarton** [comment]
I started such a library a few days ago that does that: https://github.com/mcarton/rust-derivative.
It's not finished, not really documented, not published on *crates.io*, but the POC is there :sweat_smile:

**Comment by alexreg** [comment]
Cool, thanks for sharing that. Is Macros 1.1 already fully implemented in nightly then? I may give it a go shortly.

**Comment by alexreg** [comment]
I'm wondering with all the recent Macro 2.0 developments whether the derive attribute macro might become a crate of its own in the future, such that it could be enhanced without changes to the actual compiler. Thoughts, @jseyfried @Manishearth?

**Comment by sfackler** [comment]
Custom derives can already be implemented outside of the compiler.

**Comment by alexreg** [comment]
@sfackler You misunderstand. I mean the workings of *the derive attribute macro itself*.

**Comment by sfackler** [comment]
How would pulling `libsyntax_ext::deriving` into a different crate make this feature easier to implement? Difficulty of implementation is not the limiting factor here in any case.

**Comment by Manishearth** [comment]
It actually is. The stage1 compiler cannot depend on proc macros IIRC. I
suppose you could still make it work with some shimming of libproc_macro
and including it directly in libsyntax.

**Comment by sfackler** [comment]
But that would mean that deriving can't be pulled out to a proc macro, not that pulling it out to a proc macro would make it easier to add extra control attributes, right?

**Comment by alexreg** [comment]
I was thinking more of maintainability and neat, elegant separation of concerns. Not sure on the details of what @Manishearth just said, but that could be an interesting point.

**Comment by Manishearth** [comment]
Yes.

**Comment by alexreg** [comment]
Is there no elegant way to get around this restriction?

**Comment by Manishearth** [comment]
I'm not 100% clear on whether this restriction will be a problem here.

@eddyb knows

**Comment by irexiz** [comment]
Kind of digging up  a buried corpse, but the crate https://crates.io/crates/derivative does that with 
`#[derivative(Debug="ignore")]`

I think the issue can be closed with this. Thank you @mcarton!

**Comment by Allen-Webb** [comment]
A way of skipping fields in #[derive(Debug)] would be particularly useful for structs that have to use PhantomData.

**Comment by Xuanwo** [comment]
The feature is also helpful for structs that have sensitive data like AWS s3 access_key_id & secret_access_key.

**Comment by MolotovCherry** [comment]
For me, the problem often arises when using 3rd party libraries that don't implement `Debug` on their types, which disallows me from using `derive(Debug)`. And sometimes all I need is just a quick way to derive it rather than manually implementing it.

**Comment by aljazerzen** [comment]
I think that the issue of deriving `Debug` for 3rd party types is a separate issue @MolotovCherry.

---

Regarding skipping fields, serde has a [similar mechanism](https://serde.rs/field-attrs.html#skip) for serialization:

```rust
#[derive(Serialize)
struct MyStruct {
    // serialized always
    name: String,

    // serialized never
    #[serde(skip)]
    aws_secret: String,

    // serialized when `Option::is_none(&my_struct.extra)` is `false`
    #[serde(skip_serializing_if = "Option::is_none")] 
    extra: Option<String>
}
```

Debug could work similarly:
```rust
#[derive(Debug)
struct MyStruct {
    // debug always
    name: String,

    // debug never
    #[debug(skip)]
    aws_secret: String,

    // serialized when `Option::is_none(&my_struct.extra)` is `false`
    #[debug(skip_if = "Option::is_none")] 
    extra: Option<String>
}
```

The API is can be a bit simpler compared to `serde`, as `serde` has separate mechanisms for skipping serialization and de-serialization.

**Comment by MolotovCherry** [comment]
@aljazerzen I think you misunderstand me. I'm talking about skipping fields _in my own types_ which derive debug, but _can't derive debug_, because a third party type doesn't.

```rs
struct MyType {
    other_field: u64,
    // this third party type didn't implement Debug, now I can't derive debug
    // but I don't care about getting the details of it, I just want to quickly debug MyType
    // the following would be nice
    #[debug(skip)]
    foo: ThirdPartyType
}
```

**Comment by aljazerzen** [comment]
Yes, I didn't understand what you meant. This is very related and a good argument for the feature.

**Comment by hagbard** [comment]
Alternatively, perhaps `#[derive(Debug)]` could have a flag or variant that just silently skips types without the `Debug` trait.
I'm hitting this a lot when trying to write code which interoperates between things like `itertools` and `ndarray`, since a lot of stuff in `itertools` demands `Debug`, and a bunch of stuff in `ndarray` doesn't provide it.

Can someone summarize the current (EOY '23) best workaround?

**Comment by tv42** [comment]
> Alternatively, perhaps `#[derive(Debug)]` could have a flag or variant that just silently skips types without the `Debug` trait.

I have several cases where my struct contains a binary byte slice with stuff that isn't interesting for debugging, or outright shouldn't be logged -- think crypto keys. Just because it implements `Debug` doesn't mean I want it in my debug output.

**Comment by Kobzol** [comment]
> Alternatively, perhaps #[derive(Debug)] could have a flag or variant that just silently skips types without the Debug trait.

Note that this is currently not possible, because deriving happens before type checking.

**Comment by aljazerzen** [comment]
> Can someone summarize the current (EOY '23) best workaround?

Use [derivative](https://mcarton.github.io/rust-derivative/latest/Debug.html) - an alternative derive for implementing `std::fmt::Debug`.

**Comment by wiseaidev** [comment]
We need this feature so much because we don't want to expose private keys and credentials when debugging.

**Comment by Kobzol** [comment]
As an alternative, you can also achieve that by wrapping sensitive data into a `Secret` newtype, e.g. using [`secrecy`](https://docs.rs/secrecy/latest/secrecy/).

**Comment by Sytten** [comment]
> > Can someone summarize the current (EOY '23) best workaround?
> 
> Use [derivative](https://mcarton.github.io/rust-derivative/latest/Debug.html) - an alternative derive for implementing `std::fmt::Debug`.

Derivative is effectively unmaintained, there are a couple alternatives but it really is needed built-in IMO

**Comment by geonnave** [comment]
> We need this feature so much because we don't want to expose private keys and credentials when debugging.

Same use case here, it would be great to have such a `skip` attribute.

**Comment by palant** [comment]
This also goes as cheating but a wrapper type isn’t exactly much code:

```rust
use std::fmt::Debug;
use std::marker::PhantomData;
use std::ops::{Deref, DerefMut};

struct NoDebug<T> {
    inner: T,
}

impl<T> Debug for NoDebug<T> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("skipped").finish()
    }
}

impl<T> From<T> for NoDebug<T> {
    fn from(value: T) -> Self {
        Self { inner: value }
    }
}

impl<T> Deref for NoDebug<T> {
    type Target = T;

    fn deref(&self) -> &Self::Target {
        &self.inner
    }
}

impl<T> DerefMut for NoDebug<T> {
    fn deref_mut(&mut self) -> &mut Self::Target {
        &mut self.inner
    }
}

impl<T> Default for NoDebug<T>
where
    T: Default,
{
    fn default() -> Self {
        Self {
            inner: T::default(),
        }
    }
}
```

Whatever fields shouldn’t appear in the debugging output can be wrapped then:

```yaml
#[derive(Debug, Default)]
struct Debuggable {
    hidden: NoDebug<PhantomData<i32>>,
    visible: i32,
}

fn main() {
    dbg!(Debuggable::default());
}
```

This prints:

```
Debuggable::default() = Debuggable {
    hidden: skipped,
    visible: 0,
}
```

[Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=05a1a6374f0543b62bdf730dd0fd4e1e)

**Comment by PWhiddy** [comment]
+1 this would be a super useful feature :)

**Comment by barries** [comment]
+ 3.

This would ease:

+ 1: debugging (skipping large, useless, or non-Debug fields)
+ 1: security compliance (skipping credentials), especially in regulated industries
+ 1: HIPPA compliance (PII, PHI).

**Comment by Qix-** [comment]
@barries If the difference between HIPAA compliance and non-compliance is a `#[derive(Debug)]` then you **_severely_** need to revisit your data security practices and threat model. Information security should not be relevant to this particular conversation whatsoever; this is not a security feature.

**Comment by barries** [comment]
Given the prevalence of using debug for logging in-flight data, I disagree.

**Comment by ant1fact** [comment]
+1, this would be great for ergonomics.

**Comment by atezet** [comment]
For people watching this ticket; the crate [`derive_more`](https://docs.rs/derive_more/latest/derive_more/derive.Debug.html) has (somewhat) recently added support for this in their 1.0 release.

**Comment by JEleniel** [comment]
Having such an attribute would help with various compliance requirements, such as HIPPA, by making it provable from the source code that the secured value *cannot* be logged, even by accident, because it is prevented at compile time.

**Comment by Kobzol** [comment]
You can get that guarantee by using different types for secret values (see the secrecy crate, which does exactly that). That's even safer, because you can't forget to add an attribute to such field.
